### PR TITLE
PokePointer ignore colliders that are not visible

### LINKED ISSUE #6634: When you want to grab a far object but hand is near interactive object, the ray turns off
## Overview
In some cases a user's hand is near a grabbable object but they are trying to grab something far away. Since the hand is near a grabbable object, the ray disappears.

## Solutions to try

- [x] If a hand is near a grabbable but not in the camera frustrum (not visible) then do not turn off the hand ray. (PR #6870)

- [x] If a hand is near a TOUCHABLE but not in the camera frustrum (not visible) then do not turn off the hand ray. (PR #7016 )

Current implementation for detecting grabbable objects

## Evaulation criteria
Open the near menu interaction scene and check if it's possible to have a near menu near the object but also to target objects at a distance.

### PR DESCRIPTION
**Pull request by julenka**
## Overview
Following up from #6870 this change additionally allows the **poke pointer** to ignore colliders that are not in the visual FOV.

Since this is a follow up to #6870 which is going into 2.3, I decided to request this go into stabilization. However, this is a risk that this could regress some touch behavior so I would be open to us moving this back to 2.4. Curious to hear other's opinions. I have verified in HandInteractionExample scene and have added tests.

## Changes
- Partial fix #6634. There is one more edge case for line pointers.
- Adds `Camera.IsInFovConeCached` as camera extension to ensure efficient checking of whether colliders are in FOV (as these computations run each frame).
- Adds field to PokePointer specifying whether to ignore colliders that are not in visual FOV. True by default.
- Add tests to verify that poke pointers correctly ignore colliders that aren't in FOV.

## Verification
- Verified that tests pass
- Tested in HandInteractionExample scene

> This optional section is a place where you can detail the specific type of verification 
> you want from reviewers. For example, if you want reviewers to checkout the PR locally
> and validate the functionality of specific scenarios, provide instructions
> on the specific scenarios and what you want verified.
>
> If there are specific areas of concern or question feel free to highlight them here so
> that reviewers can watch out for those issues.
>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs [hunk 1]
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
             Vector3 screenPoint = camera.WorldToViewportPoint(position);
 
             return screenPoint.z >= camera.nearClipPlane && screenPoint.z <= camera.farClipPlane
-                && screenPoint.x >= 0 && screenPoint.x <= 1 
+                && screenPoint.x >= 0 && screenPoint.x <= 1
                 && screenPoint.y >= 0 && screenPoint.y <= 1;
         }
 
+
+        // Help to clear caches when new frame runs
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: I'm curious to hear opinions on naming for these static variables, as well as whether people have ideas for how to implement this in a better way.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs [hunk 2]
```diff
@@ -51,6 +52,25 @@ public class PokePointer : BaseControllerPointer, IMixedRealityNearPointer
...
+            "radius equal to half display height.")]
+        private bool ignoreCollidersNotInFOV = true;
+       
+        /// <summary>
+        /// Whether to ignore colliders that may be near the pointer, but not actually in the visual FOV.
+        /// This can prevent accidental touches, and will allow hand rays to turn on when you may be near 
+        /// a touchable but cannot see it. Visual FOV is defined by cone centered about display center, 
+        /// radius equal to half display height.
+        /// </summary>
+        public bool IgnoreCollidersNotInFOV
```
[the remark this answers is not in the document]
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: Maybe? I'm hesitant to modify interfaces since it's hard to undo or change interfaces.

[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: I agree with Julia, feels more like feature creep for the IMixedRealityPointer interface. 

But it does seem like this would be good in a base class for PokePointer and SpherePointer

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs [hunk 3]
```diff
@@ -100,52 +100,76 @@ public IEnumerator TestLinePointers()
...
+                Assert.IsTrue(myPointer.IsInteractionEnabled, $"Pointer {myPointer.PointerName} should be enabled.");
+                spherePointer.IgnoreCollidersNotInFOV = true;
+            }
+            else if (myPointer is PokePointer pokePointer)
+            {
+                pokePointer.IgnoreCollidersNotInFOV = false;
+                yield return PlayModeTestUtilities.WaitForInputSystemUpdate();
+                Assert.IsTrue(myPointer.IsInteractionEnabled, $"Pointer {myPointer.PointerName} should be enabled.");
+                pokePointer.IgnoreCollidersNotInFOV = true;
+            }
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: This would also be good to simplify with base class shared between the two

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs (hunk 1 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: I feel like this is complicated enough to be in a utility class by itself?

Other thing that comes up is concurrency. I think....we are fine because Unity restricts all calls to the main thread and this is an extension of Camera. So this should force all calls to IsInFOVConeCached to be serialized and thus not possible for two threads to be accessing the static resources concurrently. Thoughts?

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs [hunk 4]
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
...
+        static private int lastCalculatedFrame = -1;
+        // Map from grabbable => is the grabbable in FOV for this frame. Cleared every frame
+        private static Dictionary<Collider, bool> colliderCache = new Dictionary<Collider, bool>();
+        // List of corners shared across all sphere pointer query instances --
+        // used to store list of corners for a bounds. Shared and static
+        // to avoid allocating memory each frame
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Remove "Sphere pointer query instances"

// List of corners of bounds box for current collider being processed in IsInFOVConeCached
// Shared and static to avoid allocating additional memory each frame

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Suggestion: I'd also rename to boundsCornerPoints to just be explicit and easier reading

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs [hunk 5]
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
...
+        public static bool IsInFOVConeCached(this Camera cam,
+            Collider myCollider)
+        {
+            if (lastCalculatedFrame != Time.frameCount)
+            {
+                colliderCache.Clear();
+                lastCalculatedFrame = Time.frameCount;
+            }
+
+            if (colliderCache.TryGetValue(myCollider, out bool result))
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: minor nit: make else if (colliderCache.Try) 
If we clear the cache, then we know there is no point in trying to test access

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs [hunk 6]
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
...
+                xMin = Mathf.Min(xMin, corner.x);
+                yMin = Mathf.Min(yMin, corner.y);
+                zMin = Mathf.Min(zMin, corner.z);
+                xMax = Mathf.Max(xMax, corner.x);
+                yMax = Mathf.Max(yMax, corner.y);
+                zMax = Mathf.Max(zMax, corner.z);
+            }
+
+            // edge case: check if camera is inside the entire bounds of the collider;
+            // Consider simplifying to myCollider.bounds.Contains(CameraCache.main.transform.position)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Seems like left over comment?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: I would prefer not to add more levels of inheritance until there's lots of signal (3 repeated classes) that a field could be re-used. Motivation is I find it really hard to navigate the code in MRTK because many fields are in many levels of base classes. Maybe I just need a better tool for navigating base classes (currently use class diagrams). I am thinking of making a third kind of near pointer, a Cone Pointer, as part of fixing #6634 (which I will probalby break up into multiple bugs). Then it woudl be appropriate to have a base class, I think.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs (hunk 3 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: Good point! I still would prefer to wait until we have more need, specifically when creating a ConePointer. I can create a bug for this and mention that following up from #7016 we should make a base class.

**Comment by julenka** [comment]
> Lastly, this change somewhat begs the question if there should a IsInFOVCached version? The cone is better performance

I'm not sure I understand -- are you suggesting creating a cached version of IsInFOV? I could make it, though not sure it is needed since nobody is using it right now.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: Very true, base classes impede performance and sometimes code maintainability.

Alternative suggestion...to just consider...because I myself am not sure it's better/good. Why not create a new interface, like IFOVChecker? 

Just a thought

[the remark this answers is not in the document]
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: I'd prefer to reduce the number of new interfaces we add. I think when I build the cone pointer I will probably make a base class or interface, but for now I'll leave it like this...keeps the code simpler

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs (hunk 1 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: I tried to think of a good location and name for this, but couldn't really think of anything. I'm thinking again this might be good to refactor once we have other code that does FOV checks where we want caching.

However, to help clarify things I prefixed the variables with inFOVCone so people understand this is related to the fov cone checks.

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs (hunk 4 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Renamed!

**Comment by julenka** [comment]
Troy found a bug -- If you have two cameras camera1 and camera2, if collider A is in camera1 FOV but not in camera2 FOV and we do the following:

```
camera1.IsInFovConeCached() // returns true
camera2.IsInFovConeCached() // will also return true 
```
`camera2.IsInFovConeCached()` should not return true, but it will because the cache is shared.

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs (hunk 5 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Done

File: Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs (hunk 6 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs]: Done

**Comment by julenka** [comment]
Hi @Troy-Ferrell I fixed the edge case you though of (multiple cameras causing incorrect collider cache hits) and added a test to verify my fix. Could you please take one more look?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Materials/HolographicButtonContentCage.mat [hunk 7]
```diff
@@ -175,7 +176,7 @@ Material:
     - _ClipPlane: {r: 0, g: 1, b: 0, a: 0}
     - _ClippingBorderColor: {r: 1, g: 0.2, b: 0, a: 1}
     - _ClippingPlaneBorderColor: {r: 1, g: 0.2, b: 0, a: 1}
-    - _Color: {r: 0.1254902, g: 0.1254902, b: 0.1254902, a: 1}
+    - _Color: {r: 0.3254902, g: 0.3254902, b: 0.3254902, a: 1}
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Materials/HolographicButtonContentCage.mat]: Just double check this was intentional

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs [hunk 8]
```diff
@@ -183,11 +203,17 @@ private bool FindClosestTouchableForLayerMask(LayerMask layerMask, out BaseNearI
...
 
+            Camera mainCam = CameraCache.Main;
             for (int i = 0; i < numColliders; ++i)
             {
-                var touchable = queryBuffer[i].GetComponent<BaseNearInteractionTouchable>();
+                var collider = queryBuffer[i];
+                var touchable = collider.GetComponent<BaseNearInteractionTouchable>();
                 if (touchable)
                 {
+                    if (ignoreCollidersNotInFOV && !mainCam.IsInFOVConeCached(collider))
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: nit: Feel free to ignore this for now. But should use IgnoreCollidersNotInFOV public property instead of private, to be consistent with access to private variables in case the getter/setter does particular code check etcs

File: Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs [hunk 9]
```diff
@@ -62,14 +111,38 @@ public void TearDown()
...
+            }
+        }
+
+        [Test]
+        public void TestIsInFOVConeCachedSecondCamera()
+        {
+            for (int i = 0; i < TestColliders.Count; i++)
+            {
+                var test = TestColliders[i];
+                Assert.AreEqual(test.ShouldBeInFOVCamera1, testCamera.IsInFOVConeCached(test.Collider), $"TestCollider[{i}] did not match");
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs]: Why is the testCamera done twice but not the second camera?

File: Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs [hunk 10]
```diff
@@ -62,14 +111,38 @@ public void TearDown()
             GameObjectExtensions.DestroyGameObject(testCamera.gameObject);
         }
 
+
+
+        [Test]
+        public void TestIsInFOVConeCached()
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs]: nit: Test comments explaining what tests do

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs [hunk 11]
```diff
@@ -100,52 +100,76 @@ public IEnumerator TestLinePointers()
             BaseEventSystem.enableDanglingHandlerDiagnostics = true;
         }
 
-        /// <summary>
-        /// Tests that sphere pointer grabs object when hand is insize a giant grabbable
-        /// </summary>
-        [UnityTest]
-        public IEnumerator TestSpherePointerInsideGrabbable()
+
+        private IEnumerator TestPointerViewOfViewHelper(IMixedRealityPointer myPointer, GameObject cube, TestHand testHand)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: Can you rename this? Not understanding viewofview

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: maybe meant FieldOfView?

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs (hunk 3 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: Since you arent' creating an interface or base class, maybe add else statement here that throws error saying test helper isn't valid for this type of pointer? To help notify people who might try to use this in future

File: Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        /// <summary>
+        /// Returns true if a collider's bounds is within the camera FOV. 
+        /// Utilizes a cache to test if this collider has been seen before and returns current frame's calculated result.
+        /// </summary>
+        /// <param name="myCollider">The collider to test</param>
+        public static bool IsInFOVConeCached(this Camera cam,
+            Collider myCollider)
+        {
+            // if the collider's size is zero, it is not visible. Return false.
+            if(myCollider.bounds.size == Vector3.zero || myCollider.transform.localScale == Vector3.zero)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs]:
```suggestion
            if (myCollider.bounds.size == Vector3.zero || myCollider.transform.localScale == Vector3.zero)
```

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs]: btw, should these be floating point margin of error comparison instead of exact comparison to zero?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Materials/HolographicButtonContentCage.mat (hunk 7 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Materials/HolographicButtonContentCage.mat]: Good catch! Will revert.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs (hunk 8 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: Fixed.

File: Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs (hunk 9 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs]: Good catch, I was testing that the cache was getting hit. There should be only one call to testCamera.IsInFOVConeCached

File: Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs (hunk 10 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs]: Done

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs (hunk 11 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: Was supposed to be TestPointerFieldOfViewHelper

Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: Was supposed to be TestPointerFieldOfViewHelper

File: Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs (hunk 12 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs]: I assumed no, since people might want to set something invisible by setting it exactly to Vector3.zero. Somebody might have a collider whose scale is very small but not zero, to represent say a point. Didn't want to ignore that.

**Comment by julenka** [comment]
Do not merge until 2.3 stabilization branch is unlocked

**Comment by davidkline-ms** [comment]
@julenka,

This PR is failing TestPoitnerFieldOfView and TestToggleEvents. Can you investigate?

**Comment by julenka** [comment]
> @julenka,
> 
> This PR is failing TestPoitnerFieldOfView and TestToggleEvents. Can you investigate?

Yes, I'm aware. I am able to reproduce locally and was investigating already. I haven't understood what's wrong yet, and haven't had time to look at this since last Wednedsay. This should not block release. I will have time to look at it today, but no guarantees I will fix it. If we are close to releasing please let me know and I'll rebase off of mrtk_development.

**Comment by julenka** [comment]
Completing, will keep an eye on MRTK_CI in case this change causes intermittent test failures.
